state.go

```diff
@@ -15,3 +22,366 @@ func Digest(st llb.State) (digest.Digest, error) {
 	dgst, _, _, _, err := st.Output().Vertex(ctx, c).Marshal(ctx, c)
 	return dgst, errtrace.Wrap(err)
 }
+
+type layerHistory struct {
+	empty bool
+	desc  string
+}
+
+func commitHistory(img *mdispec.DockerOCIImage, commit layerHistory) {
+	img.History = append(img.History, ocispec.History{
+		// Set a zero value on Created for more reproducible builds
+		Created:    &time.Time{},
+		CreatedBy:  commit.desc,
+		Comment:    "llblib.v0",
+		EmptyLayer: commit.empty,
+	})
+	// Set a zero value on Created for more reproducible builds
+	img.Created = &time.Time{}
+}
+
+// Merge is similar to llb.Merge but also commits history to the image config.
+func Merge(states []llb.State, opts ...llb.ConstraintsOpt) llb.State {
+	if len(states) == 1 {
+		return states[0]
+	}
+	return llb.Merge(states, opts...).Async(func(ctx context.Context, st llb.State, c *llb.Constraints) (llb.State, error) {
+		// if any of the merged states has an image config, then preserve
+		// the first one we see.
+		for _, ms := range states {
+			cs, err := loadImageConfigState(ctx, ms)
+			if err != nil {
+				continue
+			}
+			if cs.config == nil {
+				continue
+			}
+			st = st.WithValue(imageConfigKey{}, cs)
+			break
+		}
+
+		msgs := []string{}
+		for _, ms := range states {
+			dgst, _, _, _, err := ms.Output().Vertex(ctx, c).Marshal(ctx, c)
+			if err != nil {
+				return llb.State{}, errtrace.Wrap(err)
+			}
+			msgs = append(msgs, dgst.Encoded()+":/")
+		}
+
+		return withImageConfigMutator(st, func(_ context.Context, img *mdispec.DockerOCIImage) error {
+			commitHistory(img, layerHistory{
+				empty: false,
+				desc:  "MERGE " + strings.Join(msgs, " "),
+			})
+			return nil
+		}), nil
+	})
+}
+
+// Diff is similar to llb.Diff but also commits history to the image config.
+func Diff(lower, upper llb.State, opts ...llb.ConstraintsOpt) llb.State {
+	return llb.Diff(lower, upper, opts...).Async(func(ctx context.Context, st llb.State, c *llb.Constraints) (llb.State, error) {
+		lowerDgst, _, _, _, err := lower.Output().Vertex(ctx, c).Marshal(ctx, c)
+		if err != nil {
+			return llb.State{}, errtrace.Wrap(err)
+		}
+		upperDgst, _, _, _, err := upper.Output().Vertex(ctx, c).Marshal(ctx, c)
+		if err != nil {
+			return llb.State{}, errtrace.Wrap(err)
+		}
+		return withImageConfigMutator(st, func(_ context.Context, img *mdispec.DockerOCIImage) error {
+			img.Author = ""
+			img.RootFS = ocispec.RootFS{}
+			img.Config = mdispec.DockerOCIImageConfig{
+				ImageConfig: ocispec.ImageConfig{
+					Labels: img.Config.Labels,
+				},
+			}
+
+			commitHistory(img, layerHistory{
+				empty: false,
+				desc:  "DIFF " + lowerDgst.Encoded() + ":/ " + upperDgst.Encoded() + ":/",
+			})
+			return nil
+		}), nil
+	})
+}
+
+// Run is similar to llb.Run but also commits history to the image config.
+func Run(st llb.State, opts ...llb.RunOption) llb.ExecState {
```

Comment:
This one is a tad annoying, I was hoping for `state.Run(..., AddHistory()).Root()` option but I could not get any RunOption implementation working that could modify the config value stored on the state. So to add config history the usage will be `llblib.Run(state, ...).Root()`

Reply:
I ended up adding any `ApplyRun` llb.StateOption helper to make the syntax better for chaining multiple Run ops together.